traversal/scanner/GraphIterator.java

```diff
@@ -176,82 +214,29 @@ private boolean computeAnswer() {
-        toTraverse.add(procedureVertex);
+    private void failed(VertexTraverser vertexTraverser) {
+        iterate(vertexTraverser.procedureVertex.dependees()).forEachRemaining(dependee -> toRevisit.add(dependee));
+        vertexTraverser.implicitDependees.forEach(toRevisit::add);
+        vertexTraverser.clear();
+        toTraverse.add(vertexTraverser.procedureVertex);
         direction = Direction.REVISIT;
     }
 
-    private boolean verifyTraversedScopes(ProcedureVertex<?, ?> procedureVertex, Set<ProcedureVertex<?, ?>> verifyFailureCauses) {
-        if (procedureVertex.id().isScoped()) {
-            Scopes.Scoped scoped = scopes.get(procedureVertex.id().asScoped().scope());
-            return verifyScopedUpToOrder(scoped, procedureVertex.order(), verifyFailureCauses);
-        } else if (procedureVertex.isThing() && !procedureVertex.asThing().inRolePlayers().isEmpty()) {
-            for (ProcedureEdge<?, ?> edge : procedureVertex.asThing().inRolePlayers()) {
-                assert edge.isRolePlayer();
-                Scopes.Scoped scoped = scopes.get(edge.asRolePlayer().scope());
-                if (!verifyScopedUpToOrder(scoped, procedureVertex.order(), verifyFailureCauses)) return false;
-            }
-        }
-        return true;
-    }
-
-    private boolean verifyScopedUpToOrder(Scopes.Scoped scoped, int order, Set<ProcedureVertex<?, ?>> verifyFailureCauses) {
-        if (!scoped.isValidUpTo(order)) {
-            scoped.scopedOrdersUpTo(order).forEach(scopedOrder -> {
-                if (scopedOrder != order) verifyFailureCauses.add(procedure.vertex(scopedOrder));
-            });
-            return false;
-        }
-        return true;
-    }
-
-    private boolean verifyEdgeLookahead(ProcedureVertex<?, ?> procedureVertex, Set<ProcedureVertex<?, ?>> verifyFailureCauses) {
-        VertexTraverser vertexTraverser = vertexTraversers.get(procedureVertex);
-        Set<ProcedureEdge<?, ?>> verified = new HashSet<>();
-        for (ProcedureEdge<?, ?> edge : procedureVertex.orderedOuts()) {
-            VertexTraverser toVertexTraverser = vertexTraversers.get(edge.to());
-            toVertexTraverser.addEdgeInput(edge, vertexTraverser.vertex());
-            verified.add(edge);
-            if (!toVertexTraverser.hasNextVertex()) {
-                verified.forEach(e -> vertexTraversers.get(e.to()).removeEdgeInput(e));
-                toVertexTraverser.edgeInputs().forEach(e -> verifyFailureCauses.add(e.from()));
-                return false;
-            }
-        }
-        return true;
-    }
-
     private void revisit(ProcedureVertex<?, ?> procedureVertex) {
-        removeFromSuccessors(procedureVertex);
-        iterate(procedureVertex.transitiveOuts()).map(ProcedureEdge::to).forEachRemaining(toTraverse::remove);
         toTraverse.add(procedureVertex);
+        vertexTraversers.get(procedureVertex).implicitDependents.forEach(implicit -> vertexTraversers.get(implicit).clear());
```

Comment:
when re-visiting a vertex to generate the next one, we need to release all the vertices/scoped roles held by the implicit dependents. So for example:
```
1) $r --RP--> $x
2) $r --RP--> $y
```
when we backtrack to the next $x, we must release the role held by $y to guarantee `$x` will find all valid $x's. This is captured by the implicit dependencies